### PR TITLE
Sheffield | 25-SDC-Nov | Sheida Shabankari | Sprint 2 |Implement linked list in python

### PR DESCRIPTION
**Pull request by sheida-shab**
- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

This PR implements a doubly linked list in Python with O(1) time complexity for all required operations.

Added a Node class to represent elements with value, next, and previous references.

Implemented push_head to insert elements at the head and return a node handle for constant-time removal.

Implemented pop_tail to remove and return the value of the tail node.

Implemented remove to delete a node from the list using its handle, without traversing the list.

Properly maintains head and tail references and updates links for all edge cases (empty list, single node, head, tail, and middle nodes).

All provided tests pass successfully.

### CODE REVIEW

File: Sprint-2/implement_linked_list/linked_list.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        if self.head is None:
+            return None
+        elif self.head==self.tail:
+            node_value=self.head.value
+            self.head=None
+            self.tail=None
+        else:
+            node_value=self.tail.value
+            previous_node=self.tail.previous   
+            self.tail=previous_node
```
Comment (cjyuan) [inline Sprint-2/implement_linked_list/linked_list.py]: Could also consider calling `remove()` to remove the tail -- less code to maintain.

File: Sprint-2/implement_linked_list/linked_list.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            self.head.previous=None
+        else:
+            #Node is in the middle of list
+            previous_node=node_to_remove.previous        
+            next_node=node_to_remove.next
+            previous_node.next=next_node
+            next_node.previous=previous_node
+        
+        node_to_remove.next=None
+        node_to_remove.previous=None
```
Comment (cjyuan) [inline Sprint-2/implement_linked_list/linked_list.py]: Why not also clearing the reference of the removed tail in `pop_tail()`?

File: Sprint-2/implement_linked_list/linked_list.py (hunk 1 above)
Comment (sheida-shab) [reply Sprint-2/implement_linked_list/linked_list.py]: Did it.

File: Sprint-2/implement_linked_list/linked_list.py (hunk 2 above)
Comment (sheida-shab) [reply Sprint-2/implement_linked_list/linked_list.py]: Did it.

File: Sprint-2/implement_linked_list/linked_list.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            return None
+        elif self.head==self.tail:
+            node_to_remove = self.head
+            node_value = node_to_remove.value
+            self.remove(node_to_remove)
+            
+        else:
+           node_to_remove = self.tail
+           node_value = node_to_remove.value
+           self.remove(node_to_remove)
```
Comment (cjyuan) [inline Sprint-2/implement_linked_list/linked_list.py]: Wouldn't this work the same as the code on line 34-42?
```
  node_value = self.tail.value
  self.remove(self.tail)
```

**Review by cjyuan — COMMENTED** [review]
Changes look good.

**Comment by cjyuan** [comment]
Just noticed you removed these two lines from `removed()`:
```
        node_to_remove.next=None
        node_to_remove.previous=None     
```

Why?

**Comment by sheida-shab** [comment]
You are right.
After the refactor, pop_tail() no longer removes nodes directly.
That responsibility was moved to remove(), so clearing the node references should also stay in remove().
I understand this now and will keep the code there.

File: Sprint-2/implement_linked_list/linked_list.py (hunk 3 above)
Comment (sheida-shab) [reply Sprint-2/implement_linked_list/linked_list.py]: I understand the point now.
remove() handles both single-node and multiple-node cases,
so I simplified pop_tail() to just call remove(self.tail).
This makes the code shorter and cleaner.

**Comment by cjyuan** [comment]
I could not find the code for clearing references in `remove()`.

**Comment by sheida-shab** [comment]
You’re right, that was an oversight on my part.
I’ve added the reference cleanup back into remove().

**Comment by cjyuan** [comment]
All good now. Well done.

**Comment by illicitonion** [comment]
Closing PR because the SDC run has finished. Feel free to re-open if you're still working on it.
